File: hostel_management_system_backend/Services/StudentPreferencesService.cs
using System.Text.Json;
using hostel_management_system_backend.Exceptions;
using Microsoft.EntityFrameworkCore;

public interface IStudentPreferencesService
{
    Task<StudentPreferenceReadDto> GetMineAsync(Guid userId, CancellationToken cancellationToken);
    Task<StudentPreferenceReadDto> UpsertMineAsync(Guid userId, StudentPreferenceUpsertDto dto, CancellationToken cancellationToken);
}

public sealed class StudentPreferencesService : IStudentPreferencesService
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);
    private static readonly string[] ValidPriorityKeys = ["price", "distance", "rating"];

    private readonly ApplicationDbContext _db;

    public StudentPreferencesService(ApplicationDbContext db)
    {
        _db = db;
    }

    public async Task<StudentPreferenceReadDto> GetMineAsync(Guid userId, CancellationToken cancellationToken)
    {
        var preference = await _db.Set<StudentPreference>()
            .AsNoTracking()
            .FirstOrDefaultAsync(x => x.UserId == userId, cancellationToken);

        if (preference is null)
            throw new NotFoundException("Student preferences not found.");

        return ToReadDto(preference);
    }

    public async Task<StudentPreferenceReadDto> UpsertMineAsync(Guid userId, StudentPreferenceUpsertDto dto, CancellationToken cancellationToken)
    {
        await ValidateInputAsync(dto, cancellationToken);

        var selectedAmenities = NormalizeAmenities(dto.SelectedAmenities);
        var priorityOrder = NormalizePriorityOrder(dto.PriorityOrder);
        var weights = ResolveWeights(dto.Weights, priorityOrder);

        var preference = await _db.Set<StudentPreference>()
            .FirstOrDefaultAsync(x => x.UserId == userId, cancellationToken);

        if (preference is null)
        {
            preference = new StudentPreference
            {
                UserId = userId,
                CreatedAt = DateTime.UtcNow,
                IsDeleted = false
            };

            await _db.Set<StudentPreference>().AddAsync(preference, cancellationToken);
        }

        preference.UniversityId = dto.UniversityId;
        preference.MinBudget = dto.MinBudget;
        preference.MaxBudget = dto.MaxBudget;
        preference.RequiredCapacity = dto.RequiredCapacity;
        preference.SelectedAmenitiesJson = JsonSerializer.Serialize(selectedAmenities, JsonOptions);
        preference.PriorityOrderJson = JsonSerializer.Serialize(priorityOrder, JsonOptions);
        preference.PriceWeight = weights.price;
        preference.DistanceWeight = weights.distance;
        preference.RatingWeight = weights.rating;
        preference.UpdatedAt = DateTime.UtcNow;

        await _db.SaveChangesAsync(cancellationToken);
        return ToReadDto(preference);
    }

    private async Task ValidateInputAsync(StudentPreferenceUpsertDto dto, CancellationToken cancellationToken)
    {
        if (dto.UniversityId == Guid.Empty)
            throw new BadRequestException("UniversityId is required.");

        if (dto.MinBudget.HasValue && dto.MinBudget.Value < 0)
            throw new BadRequestException("MinBudget cannot be negative.");

        if (dto.MaxBudget.HasValue && dto.MaxBudget.Value < 0)
            throw new BadRequestException("MaxBudget cannot be negative.");

        if (dto.MinBudget.HasValue && dto.MaxBudget.HasValue && dto.MinBudget.Value > dto.MaxBudget.Value)
            throw new BadRequestException("MinBudget cannot be greater than MaxBudget.");

        if (dto.RequiredCapacity.HasValue && dto.RequiredCapacity.Value <= 0)
            throw new BadRequestException("RequiredCapacity must be greater than zero.");

        var universityExists = await _db.Universities
            .AsNoTracking()
            .AnyAsync(x => x.Id == dto.UniversityId, cancellationToken);

        if (!universityExists)
            throw new NotFoundException("University not found.");

        if (dto.SelectedAmenities is { Count: > 0 })
        {
            var normalized = NormalizeAmenities(dto.SelectedAmenities);
            var existingCount = await _db.Amenities
                .AsNoTracking()
                .CountAsync(a => normalized.Contains(a.Name), cancellationToken);

            if (existingCount != normalized.Count)
                throw new BadRequestException("One or more selected amenities are invalid.");
        }
    }

    private static List<string> NormalizeAmenities(List<string>? selectedAmenities)
    {
        if (selectedAmenities is null || selectedAmenities.Count == 0)
            return [];

        return selectedAmenities
            .Where(x => !string.IsNullOrWhiteSpace(x))
            .Select(x => x.Trim())
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .ToList();
    }

    private static List<string> NormalizePriorityOrder(List<string>? priorityOrder)
    {
        if (priorityOrder is null || priorityOrder.Count == 0)
            return ["price", "distance", "rating"];

        var normalized = priorityOrder
            .Where(x => !string.IsNullOrWhiteSpace(x))
            .Select(x => x.Trim().ToLowerInvariant())
            .Distinct()
            .ToList();

        if (normalized.Count != ValidPriorityKeys.Length ||
            normalized.Except(ValidPriorityKeys, StringComparer.Ordinal).Any())
        {
            throw new BadRequestException("PriorityOrder must contain exactly: price, distance, rating.");
        }

        return normalized;
    }

    private static (double price, double distance, double rating) ResolveWeights(StudentPreferenceWeightsDto? input, List<string> priorityOrder)
    {
        if (input is null)
        {
            var generated = new Dictionary<string, double>(StringComparer.Ordinal)
            {
                [priorityOrder[0]] = 0.5,
                [priorityOrder[1]] = 0.3,
                [priorityOrder[2]] = 0.2
            };

            return (generated["price"], generated["distance"], generated["rating"]);
        }

        if (input.Price < 0 || input.Distance < 0 || input.Rating < 0)
            throw new BadRequestException("Weights cannot be negative.");

        var sum = input.Price + input.Distance + input.Rating;
        if (sum <= 0)
            throw new BadRequestException("At least one weight must be greater than zero.");

        return (input.Price / sum, input.Distance / sum, input.Rating / sum);
    }

    private static StudentPreferenceReadDto ToReadDto(StudentPreference preference)
    {
        var selectedAmenities = DeserializeStringList(preference.SelectedAmenitiesJson);
        var priorityOrder = DeserializeStringList(preference.PriorityOrderJson);

        if (priorityOrder.Count == 0)
            priorityOrder = ["price", "distance", "rating"];

        return new StudentPreferenceReadDto(
            preference.UserId,
            preference.UniversityId,
            preference.MinBudget,
            preference.MaxBudget,
            preference.RequiredCapacity,
            selectedAmenities,
            priorityOrder,
            new StudentPreferenceWeightsDto(preference.PriceWeight, preference.DistanceWeight, preference.RatingWeight),
            preference.CreatedAt,
            preference.UpdatedAt);
    }

    private static List<string> DeserializeStringList(string? json)
    {
        if (string.IsNullOrWhiteSpace(json))
            return [];

        try
        {
            return JsonSerializer.Deserialize<List<string>>(json, JsonOptions) ?? [];
        }
        catch (JsonException)
        {
            return [];
        }
    }
}

File: src/student-preferences/student-preferences.service.ts
import { Injectable } from '@nestjs/common';

import { PRIORITY_KEYS } from '../common/enums/app.enums';
import {
  AppBadRequestException,
  AppNotFoundException,
} from '../common/exceptions/app-exception';
import { decimalToNumber, normalizeStringList, parseJsonStringArray } from '../common/utils/database.util';
import { DatabaseService } from '../database/database.service';
import {
  StudentPreferenceReadDto,
  StudentPreferenceUpsertDto,
  StudentPreferenceWeightsDto,
} from './dto/student-preferences.dto';

@Injectable()
export class StudentPreferencesService {
  constructor(private readonly db: DatabaseService) {}

  async getMine(userId: string): Promise<StudentPreferenceReadDto> {
    const preference = await this.db.studentPreference.findFirst({
      where: {
        userId,
        isDeleted: false,
      },
    });

    if (!preference) {
      throw new AppNotFoundException('Student preferences not found.');
    }

    return this.toReadDto(preference);
  }

  async upsertMine(userId: string, dto: StudentPreferenceUpsertDto): Promise<StudentPreferenceReadDto> {
    await this.validateInput(dto);

    const selectedAmenities = normalizeStringList(dto.selectedAmenities);
    const priorityOrder = this.normalizePriorityOrder(dto.priorityOrder);
    const weights = this.resolveWeights(dto.weights, priorityOrder);

    const existing = await this.db.studentPreference.findFirst({
      where: {
        userId,
      },
    });

    const preference = existing
      ? await this.db.studentPreference.update({
          where: { id: existing.id },
          data: {
            universityId: dto.universityId,
            minBudget: dto.minBudget ?? null,
            maxBudget: dto.maxBudget ?? null,
            requiredCapacity: dto.requiredCapacity ?? null,
            selectedAmenitiesJson: JSON.stringify(selectedAmenities),
            priorityOrderJson: JSON.stringify(priorityOrder),
            priceWeight: weights.price,
            distanceWeight: weights.distance,
            ratingWeight: weights.rating,
            updatedAt: new Date(),
            isDeleted: false,
            deletedAt: null,
          },
        })
      : await this.db.studentPreference.create({
          data: {
            userId,
            universityId: dto.universityId,
            minBudget: dto.minBudget ?? null,
            maxBudget: dto.maxBudget ?? null,
            requiredCapacity: dto.requiredCapacity ?? null,
            selectedAmenitiesJson: JSON.stringify(selectedAmenities),
            priorityOrderJson: JSON.stringify(priorityOrder),
            priceWeight: weights.price,
            distanceWeight: weights.distance,
            ratingWeight: weights.rating,
            createdAt: new Date(),
            updatedAt: new Date(),
            isDeleted: false,
          },
        });

    return this.toReadDto(preference);
  }

  private async validateInput(dto: StudentPreferenceUpsertDto) {
    if (!dto.universityId) {
      throw new AppBadRequestException('UniversityId is required.');
    }

    if (dto.minBudget !== undefined && dto.minBudget !== null && dto.minBudget < 0) {
      throw new AppBadRequestException('MinBudget cannot be negative.');
    }

    if (dto.maxBudget !== undefined && dto.maxBudget !== null && dto.maxBudget < 0) {
      throw new AppBadRequestException('MaxBudget cannot be negative.');
    }

    if (
      dto.minBudget !== undefined &&
      dto.minBudget !== null &&
      dto.maxBudget !== undefined &&
      dto.maxBudget !== null &&
      dto.minBudget > dto.maxBudget
    ) {
      throw new AppBadRequestException('MinBudget cannot be greater than MaxBudget.');
    }

    if (dto.requiredCapacity !== undefined && dto.requiredCapacity !== null && dto.requiredCapacity <= 0) {
      throw new AppBadRequestException('RequiredCapacity must be greater than zero.');
    }

    const universityExists = await this.db.university.findFirst({
      where: {
        id: dto.universityId,
        isDeleted: false,
      },
      select: { id: true },
    });

    if (!universityExists) {
      throw new AppNotFoundException('University not found.');
    }

    const selectedAmenities = normalizeStringList(dto.selectedAmenities);
    if (selectedAmenities.length > 0) {
      const existingAmenities = await this.db.amenity.findMany({
        where: {
          isDeleted: false,
          name: { in: selectedAmenities },
        },
        select: { id: true },
      });

      if (existingAmenities.length !== selectedAmenities.length) {
        throw new AppBadRequestException('One or more selected amenities are invalid.');
      }
    }
  }

  private normalizePriorityOrder(priorityOrder?: string[] | null) {
    if (!priorityOrder?.length) {
      return [...PRIORITY_KEYS];
    }

    const normalized = [...new Set(priorityOrder.map((value) => value.trim().toLowerCase()).filter(Boolean))];
    if (normalized.length !== PRIORITY_KEYS.length || normalized.some((value) => !PRIORITY_KEYS.includes(value as never))) {
      throw new AppBadRequestException('PriorityOrder must contain exactly: price, distance, rating.');
    }

    return normalized;
  }

  private resolveWeights(weights: StudentPreferenceWeightsDto | null | undefined, priorityOrder: string[]) {
    if (!weights) {
      const generated: StudentPreferenceWeightsDto = {
        price: 0,
        distance: 0,
        rating: 0,
      };
      generated[priorityOrder[0]] = 0.5;
      generated[priorityOrder[1]] = 0.3;
      generated[priorityOrder[2]] = 0.2;
      return generated;
    }

    if (weights.price < 0 || weights.distance < 0 || weights.rating < 0) {
      throw new AppBadRequestException('Weights cannot be negative.');
    }

    const sum = weights.price + weights.distance + weights.rating;
    if (sum <= 0) {
      throw new AppBadRequestException('At least one weight must be greater than zero.');
    }

    return {
      price: weights.price / sum,
      distance: weights.distance / sum,
      rating: weights.rating / sum,
    };
  }

  private toReadDto(preference: any): StudentPreferenceReadDto {
    const selectedAmenities = parseJsonStringArray(preference.selectedAmenitiesJson);
    const priorityOrder = parseJsonStringArray(preference.priorityOrderJson);

    return {
      userId: preference.userId,
      universityId: preference.universityId,
      minBudget: decimalToNumber(preference.minBudget),
      maxBudget: decimalToNumber(preference.maxBudget),
      requiredCapacity: preference.requiredCapacity,
      selectedAmenities,
      priorityOrder: priorityOrder.length ? priorityOrder : [...PRIORITY_KEYS],
      weights: {
        price: preference.priceWeight,
        distance: preference.distanceWeight,
        rating: preference.ratingWeight,
      },
      createdAt: preference.createdAt,
      updatedAt: preference.updatedAt,
    };
  }
}
